test: add unit tests

BumbleFlash opened the pull request:
PR to add unit tests to Open Trip Planner's API:-

- Plan 
- Bike Rental
- Server Info

This will be tested across both Chicago and Tampa's GTFS feed. 

"Closes #8"

BumbleFlash committed:
test: added plan.json

BumbleFlash committed:
test: added bike_rental.json

BumbleFlash committed:
test: added server_info.json

BumbleFlash committed:
feat: modified RequestParameters to be optional

BumbleFlash committed:
tests: added instrumented tests for Plan API.

BumbleFlash committed:
test: added server info api instrumented tests

BumbleFlash committed:
test: added bike rental API instrumented tests

barbeau commented on app/src/androidTest/test_resources/bike_rental.json:
```diff
@@ -0,0 +1,304 @@
+{
```
On Android, the typical way to bundle these types of raw file resources are under `androidTest/res/raw` - let's use that directory instead.

barbeau replied:
Actually, let's just move these to a JVM build `test/resources` instead - see https://github.com/CUTR-at-USF/opentripplanner-client-library/pull/11/files#r607933744.

barbeau commented on library/src/commonMain/kotlin/edu/usf/cutr/otp/plan/api/PlanApi.kt:
```diff
@@ -32,7 +32,7 @@ import kotlinx.serialization.json.Json
 
 class PlanApi(private val url: String,
-              private val requestParameters: RequestParameters) {
+              private val requestParameters: RequestParameters? = null) {
```
Doesn't seem like this should be nullable - you can't plan a trip without passing in parameters like origin and destination.

BumbleFlash replied:
I added the parameter for the instrumented tests since we don't need the origin and destination. I can't pass random values either without having to append them to the GitHub url.

barbeau replied:
Does it hurt to append them to the GitHub URL?

I'd prefer to find another solution for testing to keep the code ideal for the production use environment (i.e. not allowing null)

barbeau replied:
I'd still prefer to keep this as non-nullable.

barbeau commented on app/src/androidTest/java/com/example/myapplication/ServerInfoApiInstrumentedTest.kt:
```diff
@@ -0,0 +1,57 @@
+import org.junit.runner.RunWith
+import kotlin.coroutines.CoroutineContext
+
+@RunWith(AndroidJUnit4::class)
```
These instrumented Android tests are nice, but they are also hard to run on CI because running an emulator on CI is hard (and slow).

Let's pause this PR and finish the JVM build (desktop Java) of this project first (https://github.com/CUTR-at-USF/opentripplanner-client-library/issues/6), and then we can come back and run these unit tests from the JVM build, which will make CI much easier to manage and much faster to complete.

barbeau replied:
After more research, it seems this may help solve the issue of running an emulator on GitHub Actions and it's used by Square and others:
https://github.com/marketplace/actions/android-emulator-runner

@BumbleFlash If you keep hitting issues getting the coroutines to work right in JVM tests feel free to try this with Android instrumented tests

barbeau commented on app/src/androidTest/java/com/example/myapplication/BikeRentalApiInstrumentedTest.kt:
```diff
@@ -0,0 +1,63 @@
+    fun callBikeRentalApi() {
+        job = Job()
+
+        val bikeRentalApi = BikeRentalApi("https://raw.githubusercontent.com/CUTR-at-USF/opentripplanner-client-library/tests/app/src/androidTest/test_resources/bike_rental.json")
```
Note that the `tests` part of these URLs will need to change to `master` prior to merging, as they won't be available in this branch after merge.

barbeau commented:
Now that #13 is merged, we can use that in the unit tests. You can create a new folder for a JVM desktop package, and create a `tests` folder within that (just like a normal Java desktop project - see https://github.com/CUTR-at-USF/pelias-client-library/tree/master/src/test. We shouldn't even need to create the desktop application itself.

There are some good basic examples for Kotlin tests in this post - https://github.com/CUTR-at-USF/pelias-client-library/tree/master/src/test.

BumbleFlash committed:
Merge branch 'main' into tests

BumbleFlash commented:
@barbeau To create a new JVM desktop package, I would still need a new `gradle` config for the folder, right? 
Also, do you think writing unit tests within android is much neater?

barbeau commented:
>To create a new JVM desktop package, I would still need a new gradle config for the folder, right?

You'll need a `build.gradle` that includes `testImplementation` for the dependencies you'll use in the tests.

>Also, do you think writing unit tests within android is much neater?

Not sure (I haven't seen how this looks with Kotlin JVM code), but running instrumented Android tests on CI is hard to configure (I haven't tackled it on GitHub Actions, but it's been a huge pain on Travis) and very slow. Even if the tests for JVM aren't as clean as Android, IMHO it's worth the reliability and speed that we'll gain on CI vs. using Android.

barbeau commented:
@BumbleFlash If it's easiest you can just create a sample Gradle project the same way you created the test JVM project (https://github.com/BumbleFlash/opentripplanner-multiplatform-test), and just leave the normal Main empty.

barbeau commented:
Actually, the above is wrong - it will look different because this project is multiplatform. I found this example multiplatform project that includes JVM tests:
https://github.com/orangy/multiplatform-lib/blob/master/build.gradle#L66

BumbleFlash commented:
@barbeau The instrumented tests are slower but the unit tests should be relatively faster? If you see the android package, we have three folders and the `test` folder is used for writing tests that will execute on the development machine (JVM) and not on an android device. 

> Actually, the above is wrong - it will look different because this project is multiplatform. I found this example multiplatform project that includes JVM tests:
https://github.com/orangy/multiplatform-lib/blob/master/build.gradle#L66

Here the `jvmMain` and `jvmTest` are a part of the shared library just like `androidMain` and `iosMain` where the dispatcher code is written. The `sourceSets `are used to help the library differentiate (when needed) between the platforms in case we need to perform platform-specific code or say the Dispatcher instance. Do you think it's"right" for us to write tests inside a source set?

BumbleFlash committed:
refactor: moved json files to res/raw

barbeau commented:
>Here the jvmMain and jvmTest are a part of the shared library just like androidMain and iosMain where the dispatcher code is written. The sourceSets are used to help the library differentiate (when needed) between the platforms in case we need to perform platform-specific code or say the Dispatcher instance. I'm not sure how "right" it is for us to write tests inside a source set.

The tests for the library should be under the library folder, which I believe is the above design. The remaining question in my mind if if they can be in the common folder or need to be in platform specific one. And I believe because we're using platform specific coroutine libraries the tests would need to be in the platform specific folder.

barbeau commented:
>If you see the android package, we have three folders and the test folder is used for writing tests that will execute on the development machine (JVM) and not on an android device.

If the android tests don't require a running device or emulator to execute then then should be the same as JVM specific tests as far as CI is concerned. So that implementation would be fine with me if it's easier than tests on the JVM platform.

BumbleFlash committed:
tests: added unit tests for the APIs on the android

BumbleFlash commented:
@barbeau Got the unit tests to work finally. I'm using `latches` to pause the main thread and wait for the execution of the coroutine to complete and update the object. 

I moved the assertions outside of the callbacks to make sure it executes on the main thread. I couldn't keep it inside unless I can magically join the suspended thread to the main thread and re-execute it. Using `latches `seemed more elegant compared to re-engineering `async` and waiting for it to complete.

barbeau commented on app/src/main/res/raw/server_info.json:
```diff
@@ -0,0 +1,20 @@
+{
```
Could you put this and other JSON under a `/chicago/server_info.json` or `/tampa/server_info.json` folder so we can easily tell the source? That will make it easier to add more cities in the future too. I think we're missing tests for the Tampa data right now in OTP? If so could you please add them?

barbeau reviewed (commented):
Glad you got this working @BumbleFlash! A quick comment now, and I'll look in more detail next week.

BumbleFlash committed:
refactor: move json files to Chicago folder

BumbleFlash committed:
refactor: move json files to jvmTest/resources folder

BumbleFlash committed:
refactor: move json files to chicago folder and fixed urls

BumbleFlash committed:
Merge branch 'main' into tests

BumbleFlash committed:
chore: added test json files for Tampa Bay

BumbleFlash committed:
refactor: moved test files to cutr package

barbeau commented:
@BumbleFlash looks like there are conflicts after merging #16 - could you resolve them?

barbeau commented on app/src/test/java/edu/usf/cutr/otp/demo/chicago/BikeRentalApiUnitTest.kt:
```diff
@@ -0,0 +1,62 @@
+package edu.usf.cutr.otp.demo.chicago
```
These tests should be moved to the `/library/src/androidTests` folder if they require Android (but it looks like they don't?), and if not they should be in the `/library/src/jvmTest` folder. Currently, if we do a command like `./gradlew :library:build` these tests won't run, because they aren't in the library project.

BumbleFlash replied:
Thanks for letting me know. The android unit tests run on the CI and if I copy these to the test folders, it should still work with the JVM.

barbeau commented on library/src/commonMain/kotlin/edu/usf/cutr/otp/plan/api/PlanApi.kt:
```diff
@@ -45,11 +45,16 @@ class PlanApi(private val url: String,
-                    parameters = parameters,
-                ).takeFrom(url).buildString()
-                val json = HttpClient().get<String>(url)
+                val urlString = if (requestParameters != null) {
```
Same on keeping `requestParameters` non-nullable - we should be able to revert this

barbeau commented on app/src/androidTest/java/com/example/myapplication/BikeRentalApiInstrumentedTest.kt:
```diff
@@ -0,0 +1,63 @@
+package com.example.myapplication
```
If these tests are duplicates of the other's that don't require a live emulator/device, let's get rid of them. I have a feeling that these tests will drift out of sync with the main library tests and cause future headaches and confusion as to why they exist, as they won't be run regularly on CI.

BumbleFlash replied:
Makes sense. I was going to remove them before we merged this because they're useless as of now.

barbeau commented on app/src/test/java/edu/usf/cutr/otp/demo/chicago/PlanApiUnitTest.kt:
```diff
@@ -0,0 +1,155 @@
+
+        planApi.getPlan(
+            success = {
+                latch.countDown()
```
For all the API tests, I would do `latch.countDown()` after assigning the variable. Otherwise, there is the potential that this thread could be suspended after `latch.countDown()` but before the variable assignment executes, which could result in randomly failing tests when the main thread resumes with the asserts before the variables are assigned.

BumbleFlash replied:
Good find. Thank you.

barbeau commented on app/src/test/java/edu/usf/cutr/otp/demo/chicago/PlanApiUnitTest.kt:
```diff
@@ -0,0 +1,155 @@
+            }
+        )
+
+        try {
```
I don't believe you need this try/catch around `latch.await()`.

barbeau reviewed (changes requested):
Thanks for getting this working @BumbleFlash! Some comments in-line.

BumbleFlash committed:
Merge branch 'main' into tests

BumbleFlash commented:
@barbeau Thanks for your review! 

> I'd still prefer to keep this as non-nullable.

I'll test to see if adding parameters to the github url affects anything.

BumbleFlash committed:
chore: converted requestParameters to be a non null parameter

BumbleFlash commented:
Pushing a premature test for Tampa bay to see if it builds on the CI. For some reason, the unit test for the Plan API throws the following error:-

```
kotlinx.serialization.json.internal.JsonDecodingException: Unexpected JSON token at offset 220: Encountered an unknown key 'stopId'.
Use 'ignoreUnknownKeys = true' in 'Json {}' builder to ignore unknown keys.
JSON input: ..... Av",
      "stopId": "1:4886",
      "stopCode": "4886",
  .....

```

It runs fine on an instrument though. 

The `stopId` exists in the model, still not sure why unit tests won't recognize them.

BumbleFlash committed:
chore: moved latches after assignment

BumbleFlash committed:
chore: removed throwable debugs

BumbleFlash committed:
fix: fixed assertion errors

BumbleFlash commented:
Yep I can confirm the unit tests are not working properly on my local system for the Tampa Bay dataset. Using the CI to test my assertions.

BumbleFlash committed:
fix: fix assertions errors

BumbleFlash committed:
fix: fix assertions errors

BumbleFlash committed:
fix: fix assertions errors

BumbleFlash committed:
fix: fix assertions errors

BumbleFlash committed:
fix: fix assertions errors

BumbleFlash committed:
tests: added unit tests for tampa's bike and serverinfo and removed try catch around latch.await()

BumbleFlash committed:
chore: deleted instrumented tests

BumbleFlash commented:
@barbeau I came across this error while I was testing assertions in the `jvmtest` folder. 

```
Exception in thread "DefaultDispatcher-worker-1 @coroutine#1" java.lang.NoClassDefFoundError: Ljava/net/http/HttpClient;
	at java.lang.Class.getDeclaredFields0(Native Method)
	at java.lang.Class.privateGetDeclaredFields(Class.java:2583)
	at java.lang.Class.getDeclaredField(Class.java:2068)
	at java.util.concurrent.atomic.AtomicLongFieldUpdater$CASUpdater$1.run(AtomicLongFieldUpdater.java:388)
	at java.util.concurrent.atomic.AtomicLongFieldUpdater$CASUpdater$1.run(AtomicLongFieldUpdater.java:386)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.util.concurrent.atomic.AtomicLongFieldUpdater$CASUpdater.<init>(AtomicLongFieldUpdater.java:385)
	at java.util.concurrent.atomic.AtomicLongFieldUpdater.newUpdater(AtomicLongFieldUpdater.java:89)
	at io.ktor.client.engine.java.JavaHttpEngine.<clinit>(JavaHttpEngine.kt)
	at io.ktor.client.engine.java.Java.create(Java.kt:16)
	at io.ktor.client.HttpClientKt.HttpClient(HttpClient.kt:42)
```

The build fails at this line: github.com/CUTR-at-USF/opentripplanner-client-library/blob/tests/library/src/commonMain/kotlin/edu/usf/cutr/otp/plan/api/PlanApi.kt#L56. 

I'm guessing the library is not being bundled correctly and the tests don't recognize the library's dependecy. Any wisdom on what's exactly going on and how to fix it?

barbeau commented:
>I came across this error while I was testing assertions in the jvmtest folder.

@BumbleFlash Is this only happening locally for you? It seems like it's passing on CI?

BumbleFlash commented:
> @BumbleFlash Is this only happening locally for you? It seems like it's passing on CI?

I haven't pushed the `jvmTest` to the PR yet. CI is only running the android unit tests. I'm happy to push it in the CI and see if it builds.

BumbleFlash committed:
test: added jvmTest for the APIs (fails locally)

barbeau commented on library/src/jvmTest/kotlin/edu/usf/cutr/otp/resources/tampa/plan.json:
```diff
@@ -0,0 +1,369 @@
+{
```
The `resources` directory and subdirectories and JSON files should be moved to `src/jvmTest/resources`. Currently it's under the source code (`kotlin`) directory. For an example see https://github.com/CUTR-at-USF/pelias-client-library/tree/master/src/test.

barbeau reviewed (commented):
One comment in-line.

> I came across this error while I was testing assertions in the jvmtest folder.

Looks like it's trying to find the Ktor HTTP engine but can't at runtime. I've poked at this a bit and can't figure out what's going on yet. I tried switching from the Java engine to the Apache engine for both `jvmMain` and `jvmTest` with:

```kotlin
implementation("io.ktor:ktor-client-apache:1.2.3-1.3.50-eap-5")
```

...instead of:

```kotlin
implementation("io.ktor:ktor-client-java:$ktorVersion")
```

...and interestingly I do see a difference - it fails like it did when the main thread wasn't being halted during the test:

```
expected:<07ed36f0149a00dcd9a91f4d225bc4bb> but was:<null>
Expected :07ed36f0149a00dcd9a91f4d225bc4bb
Actual   :null
<Click to see difference>

java.lang.AssertionError: expected:<07ed36f0149a00dcd9a91f4d225bc4bb> but was:<null>
	at org.junit.Assert.fail(Assert.java:89)
	at org.junit.Assert.failNotEquals(Assert.java:835)
	at org.junit.Assert.assertEquals(Assert.java:120)
	at org.junit.Assert.assertEquals(Assert.java:146)
	at edu.usf.cutr.otp.chicago.BikeRentalApiUnitTest.testBikeRentalApi(BikeRentalApiUnitTest.kt:32)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.runTestClass(JUnitTestClassExecutor.java:110)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:58)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:38)
	at org.gradle.api.internal.tasks.testing.junit.AbstractJUnitTestClassProcessor.processTestClass(AbstractJUnitTestClassProcessor.java:62)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.processTestClass(SuiteTestClassProcessor.java:51)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at com.sun.proxy.$Proxy2.processTestClass(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.processTestClass(TestWorker.java:119)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:182)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:164)
	at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:414)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
	at java.lang.Thread.run(Thread.java:748)
```

barbeau commented:
I get similar results when trying Ktor Jetty engine:
```kotlin
                implementation("io.ktor:ktor-client-jetty:1.5.3")
```

```
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.

expected:<07ed36f0149a00dcd9a91f4d225bc4bb> but was:<null>
Expected :07ed36f0149a00dcd9a91f4d225bc4bb
Actual   :null
<Click to see difference>

java.lang.AssertionError: expected:<07ed36f0149a00dcd9a91f4d225bc4bb> but was:<null>
	at org.junit.Assert.fail(Assert.java:89)
	at org.junit.Assert.failNotEquals(Assert.java:835)
	at org.junit.Assert.assertEquals(Assert.java:120)
	at org.junit.Assert.assertEquals(Assert.java:146)
	at edu.usf.cutr.otp.chicago.BikeRentalApiUnitTest.testBikeRentalApi(BikeRentalApiUnitTest.kt:32)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.runTestClass(JUnitTestClassExecutor.java:110)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:58)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:38)
	at org.gradle.api.internal.tasks.testing.junit.AbstractJUnitTestClassProcessor.processTestClass(AbstractJUnitTestClassProcessor.java:62)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.processTestClass(SuiteTestClassProcessor.java:51)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at com.sun.proxy.$Proxy2.processTestClass(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.processTestClass(TestWorker.java:119)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:182)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:164)
	at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:414)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
	at java.lang.Thread.run(Thread.java:748)

```

BumbleFlash commented:
@barbeau Thanks for testing this for me. The null assertions are the result of any failed API calls or errors while parsing the json object. I'll let you know once I test it on my end.

BumbleFlash committed:
test: moved resources out of kotlin

BumbleFlash committed:
test: add json tests

BumbleFlash committed:
test: fixed urls in unit tests

BumbleFlash commented:
@barbeau So when I used Jetty engines, a `java.lang.IllegalStateException: No Client ALPNProcessors!` was thrown. Looks like ALPNs are only available after JDK9. 

I switched to `implementation("io.ktor:ktor-client-apache:1.2.3-1.3.50-eap-5")` but it wouldn't parse my JSON without explicitly specifying content headers. 

I updated the library to `implementation("io.ktor:ktor-client-apache:1.5.0")`, it worked! Yay.

BumbleFlash commented:
Cancelled the previous workflow: https://github.com/CUTR-at-USF/opentripplanner-client-library/pull/11/checks?check_run_id=2343263914. It was waiting indefinitely for the tests to pass.

BumbleFlash committed:
build: switched to apache and removed redundant dependencies.

barbeau commented:
>I updated the library to implementation("io.ktor:ktor-client-apache:1.5.0"), it worked! Yay.

Nice! 🚀

BumbleFlash commented:
@barbeau Tested the multiplatform-test project and it seems to work perfectly with the client engine change. It's ready for review now. :)

barbeau commented:
So given that there is definitely different behaviors on different engines (including JVM vs. Android), I'm wondering if we should try and keep the tests that also run on Android (but don't require emulator or device).

Could we move the assert statements to a method in the common section, and then just call that from both androidTest and jvmTest?

It wouldn't be much code duplication if we didn't need all the asserts twice (one in androidTest and one in jvmTest).

**EDIT** Which reminds me, all the tests in `app` should move to `androidTest`.

barbeau commented:
~~If we do the above, we should also only have one set of JSON files, in the jvmTest resource directory.~~ Actually, given that these are used across platforms, they probably make more sense in the `commonTest` directory.

barbeau commented:
Actually, I think all the test code within the methods are the same? Maybe all the code within the existing methods could move to methods in common that are invoked from androidTest and jvmTest?

Normally this code would live in commonTest by itself, but my understanding is that we can't do that here because Ktor needs an HTTP engine, and those only exist on each platform, not in common.

BumbleFlash commented:
@barbeau Yes, if we move the tests to `commontests` folder, I don't think it will have access to any client engines. As you said, the engines are not platform-independent and the API functions should be explicitly called from the respective platforms. But don't the asserts have to be in the platform code if we are testing how the library works in a project?

barbeau commented:
>But don't the asserts have to be in the platform code if we are testing how the library works in a project?

I'm not entirely sure about this, but I would think that if you put the code in commonTests and you treat it like a library API call from the platform folders (e.g., jvmTest or androidTest) the code would still run on the platform JAR/thread, which would use the platform engine.

barbeau commented:
For example, see https://vkuzel.com/shared-test-sources-in-gradle-multi-module-project.

There is a link to this article in this SO post - https://stackoverflow.com/questions/34602954/gradle-add-dependency-to-tests-of-another-module.

barbeau commented:
@BumbleFlash After looking at this myself I'm not sure the shared test code solution is straightforward, so I would propose we move the existing tests in `app` to `androidTest`, and otherwise leave everything else as-is. We can tackle the shared code in a separate PR.

BumbleFlash committed:
refactor: moved all the code from android to androidTest inside library.

BumbleFlash commented:
@barbeau moved the code to `androidTest `and the build will run two sets of tests. Looks like it's passing all the tests.

barbeau reviewed (approved):
Awesome, thanks @BumbleFlash! 🚀

barbeau commented:
I'll update the JSON file URLs in the master branch so the tests pass.

BumbleFlash commented:
Yep opening a PR for that.

barbeau commented:
Ok - feel free if you've got it ready

BumbleFlash commented:
Just saw the "I will" in your message. Sorry about that. I'm happy to do it anyways!

barbeau commented:
No worries, thanks for the PR!
